### PR TITLE
Param conv

### PR DESCRIPTION
**Pull request by paulidale**
Relax the restrictions as to the size of integers that can be automatically coerced by the param helpers.

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/params.c [hunk 1]
```diff
@@ -62,26 +63,173 @@ void OSSL_PARAM_set_all_unmodified(OSSL_PARAM *p)
             p++->return_size = OSSL_PARAM_UNMODIFIED;
 }
 
+/* Return non-zero if the signed number is negative */
+static int is_negative(const void *number, size_t s)
+{
+    const unsigned char *n = number;
+    DECLARE_IS_ENDIAN;
+
+    return 0x80 & (IS_BIG_ENDIAN ? *n : n[s - 1]);
```
Comment (levitte) [inline crypto/params.c:72]: Nit: perhaps `n[0]` instead of `*n`, for code consistency on this line?

File: crypto/params.c [hunk 2]
```diff
@@ -62,26 +63,173 @@ void OSSL_PARAM_set_all_unmodified(OSSL_PARAM *p)
...
+
+    if (IS_BIG_ENDIAN) {
+        if (src_len < dest_len) {
+            n = dest_len - src_len;
+            memset(dest, pad, n);
+            memcpy(dest + n, src, src_len);
+        } else {
+            n = src_len - dest_len;
+            if (!check_sign_bytes(src, n, pad)
+                    || (signed_int && ((pad ^ src[n]) & 0x80) != 0))
```
Comment (levitte) [inline crypto/params.c:106]: I'm not sure I understand what this expression does...  would you mind explaining it?

Comment (levitte) [reply crypto/params.c:106]: Oh never mind, now I get it, I didn't notice that you implemented signed\<-\>unsigned coersion!

**Review by levitte — APPROVED** [review]

File: crypto/params.c (hunk 1 above)
Comment (paulidale) [reply crypto/params.c:72]: Not a huge fan but I'll change it.

File: crypto/params.c (hunk 2 above)
Comment (paulidale) [reply crypto/params.c:110]: I'll add a comment, this actually has nothing to do with signed<->unsigned conversions.

E.g. when converting the sixteen bit signed integer -253 to an eight bit signed value: FF03 truncates to 03 and passes the `check_sign_bytes` condition.

**Comment by paulidale** [comment]
Merged to master.  Thanks for the comments.

An aside: should we consider removal of the size based conditioning?  That code is effectively a fast path and shouldn't change the functionality.  For small memory devices, I've conditioned it out but doing this always will reduce complexity.

**Pull request closed without merging**